### PR TITLE
Add block iothread test cases

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:2446,2447,2448

### CODE REVIEW

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=block_iothread_test --iothread_scheme=roundrobin --nr_iothreads=2 --platform=x86_64 --guestname=Win2019 --driveformat=virtio_blk  --imageformat=qcow2 --machines=q35 --customsparams="vm_mem_limit = 12G\nimage_aio=native" --firmware=default_bios  --netdst=virbr0 --clone=no

 (1/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.legacy_cfg.q35: STARTED
 (1/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.legacy_cfg.q35: PASS (126.36 s)
 (2/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_cfg.q35: STARTED
 (2/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_cfg.q35: FAIL: Get unexpeced stg1 iothread list:{'t2', 't3'} set() (121.71 s)
 (3/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_rr.with_rounds.q35: STARTED
 (3/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_rr.with_rounds.q35: FAIL: Get unexpeced stg1 iothread len:2 0 (124.81 s)
 (4/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_rr.with_no_round.q35: STARTED
 (4/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_rr.with_no_round.q35: FAIL: Get unexpeced stg1 iothread len:1 0 (122.69 s)
 (5/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_full.q35: STARTED
 (5/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_full.q35: FAIL: Get unexpeced stg1 iothread len:4 0 (122.64 s)
 (6/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.miss_vq.q35: STARTED
 (6/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.miss_vq.q35: PASS (7.01 s)
 (7/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.over_vq.q35: STARTED
 (7/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.over_vq.q35: PASS (7.02 s)
 (8/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.overlap_vq.q35: STARTED
 (8/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.overlap_vq.q35: PASS (7.02 s)
 (9/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.miss_vqs.q35: STARTED
 (9/9) Host_RHEL.m9.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.miss_vqs.q35: PASS (7.01 s)

(Failed case is expected due to [RHEL-39948]

**Comment by qingwangrh** [comment]
@XueqiangWei @zhenyzha @menli820  Please help to review,thanks.

**Comment by menli820** [comment]
LGTM

```
 (02/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.legacy_cfg.q35: STARTED
 (02/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.legacy_cfg.q35: PASS (142.35 s)
 (03/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_cfg.q35: STARTED
 (03/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_cfg.q35: FAIL: Get unexpeced stg1 iothread list:{'t2', 't3'} set() (138.07 s)
 (04/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_rr.with_rounds.q35: STARTED
 (04/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_rr.with_rounds.q35: FAIL: Get unexpeced stg1 iothread len:2 0 (169.23 s)
 (05/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_rr.with_no_round.q35: STARTED
 (05/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_rr.with_no_round.q35: FAIL: Get unexpeced stg1 iothread len:1 0 (168.66 s)
 (06/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_full.q35: STARTED
 (06/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_cfg.vq_full.q35: FAIL: Get unexpeced stg1 iothread len:4 0 (164.46 s)
 (07/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.miss_vq.q35: STARTED
 (07/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.miss_vq.q35: PASS (9.17 s)
 (08/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.over_vq.q35: STARTED
 (08/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.over_vq.q35: PASS (9.10 s)
 (09/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.overlap_vq.q35: STARTED
 (09/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.overlap_vq.q35: PASS (9.10 s)
 (10/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.miss_vqs.q35: STARTED
 (10/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.block_iothread_test.with_negative.miss_vqs.q35: PASS (9.03 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 5 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

**Review by menli820 — APPROVED** [review]

**Comment by XueqiangWei** [comment]
@zhenyzha Could you help take a look at it? Many thanks.

**Comment by zhenyzha** [comment]
` (01/10) unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.arm64-pci: PASS (1625.08 s)
 (02/10) block_iothread_test.with_cfg.legacy_cfg.arm64-pci: PASS (161.51 s)
 (03/10) block_iothread_test.with_cfg.vq_cfg.arm64-pci: FAIL: Get unexpeced stg1 iothread list:{'t2', 't3'} set() (160.10 s)
 (04/10) block_iothread_test.with_cfg.vq_rr.with_rounds.arm64-pci: FAIL: Get unexpeced stg1 iothread len:2 0 (158.23 s)
 (05/10) block_iothread_test.with_cfg.vq_rr.with_no_round.arm64-pci: FAIL: Get unexpeced stg1 iothread len:1 0 (154.30 s)
 (06/10) block_iothread_test.with_cfg.vq_full.arm64-pci: FAIL: Get unexpeced stg1 iothread len:4 0 (156.60 s)
 (07/10) block_iothread_test.with_negative.miss_vq.arm64-pci: PASS (14.35 s)
 (08/10) block_iothread_test.with_negative.over_vq.arm64-pci: PASS (14.35 s)
 (09/10) block_iothread_test.with_negative.overlap_vq.arm64-pci: PASS (14.37 s)
 (10/10) block_iothread_test.with_negative.miss_vqs.arm64-pci: PASS (14.28 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 4 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0`

LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

**Review by zhenyzha — APPROVED** [review]

File: qemu/tests/block_iothread_test.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            cmd = cmd.format(output_path)
+        else:
+            output_path = get_linux_drive_path(session, img)
+            cmd = guest_cmd.format(output_path)
+
+        logger.debug(cmd)
+        session.cmd(cmd)
+
+    def check_image_iothread():
+        images = params.get("check_iothread_images", "").strip().split()
```
Comment (XueqiangWei) [inline qemu/tests/block_iothread_test.py:59]: The following method can also work.
images = params.get_list("check_iothread_images")

File: qemu/tests/block_iothread_test.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    vm = None
+    session = None
+    expect_to_fail = params.get("expect_to_fail", "no")
+    err_msg = params.get("err_msg", "undefined unknown error")
+    start_vm = params.get("start_vm")
+    try:
+
+        timeout = params.get_numeric("login_timeout", 180)
+        test_images = params["test_images"].split()
```
Comment (XueqiangWei) [inline qemu/tests/block_iothread_test.py:114]: test_images = params.get_list("test_images") also work.

File: qemu/tests/block_iothread_test.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        locals_var = locals()
+        if guest_operation:
+            logger.debug("Execute guest operation %s" % guest_operation)
+            locals_var[guest_operation]()
+
+        logger.debug("Destroy VM...")
+        session.close()
+        vm.destroy()
+        vm = None
+    except (virt_vm.VMCreateError, ShellCmdError) as e:
```
Comment (XueqiangWei) [inline qemu/tests/block_iothread_test.py:138]: Suggest adding virt_vm.VMStartError as well, we may hit this issue. Thanks.
(virt_vm.VMCreateError, virt_vm.VMStartError, ShellCmdError)

**Comment by XueqiangWei** [comment]
commit signature verification has been enabled, please re-push with -S. Thanks.

File: qemu/tests/block_iothread_test.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            cmd = cmd.format(output_path)
+        else:
+            output_path = get_linux_drive_path(session, img)
+            cmd = guest_cmd.format(output_path)
+
+        logger.debug(cmd)
+        session.cmd(cmd)
+
+    def check_image_iothread():
+        images = params.get_list("check_iothread_images", "")
```
Comment (XueqiangWei) [inline qemu/tests/block_iothread_test.py:59]: images = params.get_list("check_iothread_images") is enough, the default value " " will be set by get_list. Thanks.

**Review by XueqiangWei — APPROVED** [review]
LGTM,
Acked-by: xuwei@redhat.com

**Pull request merged**